presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java

```diff
@@ -229,10 +238,27 @@ public void start()
             }, 1, 1, MILLISECONDS);
             if (isResourceManagerEnabled) {
                 resourceGroupRuntimeExecutor.start();
+                resourceGroupHeartbeatExecutor.start();
             }
         }
     }
 
+    private void sendResourceGroupHeartbeat()
+    {
+        ImmutableList.Builder<ResourceGroupRuntimeInfo> resourceGroupRuntimeInfos = ImmutableList.builder();
+        for (InternalResourceGroup resourceGroup : rootGroups) {
```

Comment:
i think what you need is to traverse the tree and if it's a leaf resource group, then only send their info as part of the heartbeat. This will only send the info for the root groups only.

Reply:
Ah you are right, the soft concurrency which is used to compute adjusted queue size in static though running/queued queries reflects the leaf nodes.